### ISSUE TITLE
[Q]: wandb irregularly can't synchronize sb3 default log metrics in the 'collect_rollout' function of the source code, and differences happened in two runnings with identical code

### ISSUE THREAD

**Issue by ZhizhuoUp** [wandb/wandb#9166]
### Ask your question

I tried to use wandbcallback in sb3 with my customized callback function. However, I found a weird result. Please click this link to access its details. I run my identical code twice but got different results.
[two runnings](https://wandb.ai/zhizhuoup/zzz-share?nw=nwuserzhizhuoup)

Here are parts of my script including the definition of the custom callback function and configuration of the logger

```python
class CustomTensorboardCallback(BaseCallback):
    def __init__(self, verbose=0):
        super(CustomTensorboardCallback, self).__init__(verbose)
        self.writer = None
        self.episode_rewards = []
        self.episode_count = 0

    def _on_training_start(self):
        # Initialize TensorBoard SummaryWriter, the directory is obtained from the current logger configuration
        self.writer = SummaryWriter(log_dir=self.logger.dir)
        if self.verbose > 0:
            print("Custom TensorBoard logging started.")

    def _on_step(self):

        # Log custom metrics at each training step
        if self.n_calls % 1000 == 0:  # record in every 1000 steps

            total_rewards = self.training_env.env.env.total_rewards
            searching_reward = self.training_env.env.env.searching_reward
            signal_reward = self.training_env.env.env.signal_reward
            angle_reward = self.training_env.env.env.angle_reward
            omega_reward = self.training_env.env.env.omega_reward
            action_reward = self.training_env.env.env.action_reward

            self.logger_record(name='total_rewards', value=total_rewards)
            self.logger_record(name='searching_reward', value=searching_reward)
            self.logger_record(name='signal_reward', value=signal_reward)
            self.logger_record(name='angle_reward', value=angle_reward)
            self.logger_record(name='omega_reward', value=omega_reward)
            self.logger_record(name='action_reward', value=action_reward)

            self.writer_add(name='total_rewards', value=total_rewards)
            self.writer_add(name='searching_reward', value=searching_reward)
            self.writer_add(name='signal_reward', value=signal_reward)
            self.writer_add(name='angle_reward', value=angle_reward)
            self.writer_add(name='omega_reward', value=omega_reward)
            self.writer_add(name='action_reward', value=action_reward)

        # Make sure to call the default logging function
        if self.logger is not None:
            # Example: Record the default number of training steps
            if self.n_calls % 1000 == 0:  # record in every 1000 step

                print('this is current num_timesteps in step', self.num_timesteps)
                self.logger.record("default_metrics/step", self.num_timesteps)
                pass

        return True

    def writer_add(self, name, value, task=None):

        self.writer.add_histogram(f"rewards/{name}", value, self.num_timesteps)
        self.writer.add_scalar(f"rewards/{name}_mean", value.mean(), self.num_timesteps)
        self.writer.add_scalar(f"rewards/{name}_std", value.std(), self.num_timesteps)

    def logger_record(self, name, value, task=None):

        self.logger.record(f"rewards/{name}", value)
        self.logger.record(f"rewards/{name}_mean", value.mean())
        self.logger.record(f"rewards/{name}_std", value.std())

    def _on_training_end(self):
        # close SummaryWriter
        if self.writer:
            self.writer.close()
        if self.verbose > 0:
            print("Custom TensorBoard logging ended.")

    def _on_rollout_end(self):

        infos = self.locals.get("infos", [])
        self.logger.record("rollout_metrics/step", self.num_timesteps)
```

```python
# configure the logger
#! this logger the same as the logger used in customized callback function!
# new_logger = configure(log_dir, ["stdout", "tensorboard"])
new_logger = configure(os.path.join(log_dir, 'logs'), ["json", "tensorboard", "stdout"])
agent.set_logger(new_logger)

#! configure the wandb
config = {"scale_searching": env_cfg.rew_scale_searching,
          "scale_signal": env_cfg.rew_scale_signal,
          "scale_angle": env_cfg.rew_scale_angle,
          "scale_omega": env_cfg.rew_scale_omega,
          "scale_action": env_cfg.rew_scale_action,
          "action_error": torque_error,
          "attitude_error": attitude_error,
          "sat": env_cfg.sat}
run = wandb.init(
    project="RL_Attitude",
    config=config,
    name=running_name,
    sync_tensorboard=True,  # auto-upload sb3's tensorboard metrics
    save_code=True,  # optional
)
wandb.define_metric("*", step_metric="global_step")
wandb_configure = {}


# callbacks for agent
checkpoint_callback = CheckpointCallback(save_freq=10000, 
                                         save_path=os.path.join(log_dir, 'models'), 
                                         name_prefix="model", 
                                         verbose=2)
custom_callback = CustomTensorboardCallback(verbose=2)
#! means saving per (save_freq * num_envs)
wandb_callback = WandbCallback(gradient_save_freq=100,
                               verbose=1)

# train the agent
agent.learn(total_timesteps=n_timesteps, 
            callback=CallbackList([checkpoint_callback, custom_callback, wandb_callback]),
            )

# save the final model
agent.save(os.path.join(log_dir, "model_last"))

# close the simulator
env.close()

run.finish()
```

I found for the running named Mark1, wandb can upload all listed metrics to the web app, but it only uploads the first five sets of data for the sb3 default metrics (including rollout/ep_len_mean, rollout/ep_rew_mean, train/actor_loss, train/critic_loss, train/learning_rate) and my customized metric such as rollout_metric/step.
![Image](https://github.com/user-attachments/assets/38d731ab-d58f-4f55-a0d5-3e323baff06f)

For the running named Mark2, wandb can't upload the metrics I mentioned that only synchronized the last five times in running Mark1. However, Mark2 can record my customized reward metrics completely and correctly, while in Mark1 these metrics show a histogram for the first two data (That is incorrect because the mean value and std value of every customized reward metrics should be a scalar in a certain step, not a histogram).
![Image](https://github.com/user-attachments/assets/d71f8a08-b796-4348-9e6e-22ee39cecd49)

I run these two running at the same time. Maybe this issue is caused by the potential wandb API conflict between two active python running instances?

I would appreciate it if anyone could provide help. I have been stuck in this issue for several days and got overwhelmed.

**Comment by thanos-wandb** [comment]
Hi @ZhizhuoUp thank you for reporting this issue and providing detailed information. Since you're using the `sync_tensorboard=True` argument, `wandb` automatically logs metrics by syncing the `tfevents` file. On each run's page, you can find a `TensorBoard` tab in the left panel.

Could you please check the TensorBoard integration for both of your runs? Do you see the metrics being captured correctly there, or does the same issue occur?

It’s possible that the issue might be related to the specified TensorBoard directory, potentially causing a conflict when two runs are instantiated simultaneously. Could you provide us with a minimal, self-contained example to reproduce the issue on our end? Additionally, how are you starting these two runs together? Looking forward to your response so we can assist you further.

**Comment by thanos-wandb** [comment]
Hi @ZhizhuoUp we wanted to follow up with you regarding your support request as we have not heard back from you. Please let us know if we can be of further assistance or if your issue has been resolved.

**Comment by thanos-wandb** [comment]
Hi @ZhizhuoUp since we have not heard back from you we are going to close this request. If you would like to re-open the conversation, please let us know!
